src/main/java/org/opensearch/knn/index/query/rescore/RescoreSearcher.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.rescore;
+
+import lombok.AllArgsConstructor;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.Bits;
+import org.opensearch.knn.index.query.KNNWeight;
+import org.opensearch.knn.index.query.ResultUtil;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+
+@AllArgsConstructor
+public final class RescoreSearcher {
+
+    private final ResultUtil resultUtil;
+
+    /**
+     * Executes a k-NN query by executing a first pass search over an index with quantized vectors with
+     * an oversample factor and then rescoring down to k results for the whole shard.
+     *
+     * @param indexSearcher The index searcher for the current search
+     * @param leafReaderContexts The leaf reader contexts for the current search
+     * @param knnWeight The knn weight for the current search
+     * @return The top k results for the shard
+     * @throws IOException If an error occurs during the search
+     */
+    public TopDocs[] search(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        int firstPassK,
+        int finalK
+    ) throws IOException {
+        List<Map<Integer, Float>> perLeafResults = firstPassSearch(indexSearcher, leafReaderContexts, knnWeight, firstPassK);
+        return rescoreResults(indexSearcher, leafReaderContexts, knnWeight, perLeafResults, finalK);
+    }
+
+    private List<Map<Integer, Float>> firstPassSearch(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        int k
+    ) throws IOException {
+        List<Callable<Map<Integer, Float>>> tasks = new ArrayList<>(leafReaderContexts.size());
+        for (LeafReaderContext leafReaderContext : leafReaderContexts) {
+            tasks.add(() -> firstPassSearchLeaf(leafReaderContext, knnWeight));
+        }
+        return resultUtil.reduceToTopK(indexSearcher.getTaskExecutor().invokeAll(tasks), k);
+    }
+
+    private Map<Integer, Float> firstPassSearchLeaf(LeafReaderContext ctx, KNNWeight queryWeight) throws IOException {
+        Map<Integer, Float> leafDocScores = queryWeight.searchLeaf(ctx);
+        if (leafDocScores.isEmpty()) {
+            return leafDocScores;
+        }
+
+        Bits liveDocs = ctx.reader().getLiveDocs();
+        return leafDocScores.entrySet()
+            .stream()
+            .filter(entry -> liveDocs == null || liveDocs.get(entry.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
> Also, why we are not adding this in the function: queryWeight.searchLeaf(ctx) at the end ?

I didnt make the change in k-NN Weight, because it would change the behavior of how we currently support non-rescored search. I figured itd be best in this change to avoid chaning behavior of that path just to reduce blast radius.

Reply:
I can add comment

Reply:
Please add a comment. I think after adding deletedFilters logic this code should go away. Ref: https://github.com/opensearch-project/k-NN/issues/1491